src/ConstantBounds.cpp

```diff
@@ -0,0 +1,1141 @@
+                gatherer.a.push_back(a[i]);
+                i++;
+            } else {
+                // j term is earlier than i term
+                gatherer.b.push_back(b[j]);
+                j++;
+            }
+        }
+
+        // Wrap up tail conditions
+        while (i < a.size()) {
+            gatherer.a.push_back(a[i]);
+            i++;
+        }
+
+        while (j < b.size()) {
+            gatherer.b.push_back(b[j]);
+            j++;
+        }
+
+        return gatherer;
+    }
+
+    template<class BinOp>
+    Expr visit_binary_op(const BinOp *op) {
+        if (op->type == Int(32)) {
+            const Expr mutate_a = mutate(op->a);
+            const Expr mutate_b = mutate(op->b);
+            std::vector<AffineTerm> a_terms = extract_and_simplify(mutate_a);
+            std::vector<AffineTerm> b_terms = extract_and_simplify(mutate_b);
+            AffineTermsGather gathered = extract_like_terms(a_terms, b_terms);
+            if (gathered.c.empty()) {
+                Expr a = reconstruct_summation(a_terms, op->type);
+                Expr b = reconstruct_summation(b_terms, op->type);
+                return BinOp::make(a, b);
+            } else {
+                Expr like_terms = reconstruct_summation(gathered.c, op->type);
+                Expr a = reconstruct_summation(gathered.a, op->type);
+                Expr b = reconstruct_summation(gathered.b, op->type);
+                return BinOp::make(a, b) + like_terms;
+            }
+        } else {
+            return IRGraphMutator::visit(op);
+        }
+    }
+
+    Expr visit(const Min *op) override {
+        return visit_binary_op<Min>(op);
+    }
+
+    Expr visit(const Max *op) override {
+        return visit_binary_op<Max>(op);
+    }
+
+    Expr visit(const Select *op) override {
+        if (op->type == Int(32)) {
+            const Expr true_value = mutate(op->true_value);
+            const Expr false_value = mutate(op->false_value);
+            std::vector<AffineTerm> a_terms = extract_and_simplify(true_value);
+            std::vector<AffineTerm> b_terms = extract_and_simplify(false_value);
+            AffineTermsGather gathered = extract_like_terms(a_terms, b_terms);
+            if (gathered.c.empty()) {
+                Expr a = reconstruct_summation(a_terms, op->type);
+                Expr b = reconstruct_summation(b_terms, op->type);
+                return Select::make(op->condition, a, b);
+            } else {
+                Expr like_terms = reconstruct_summation(gathered.c, op->type);
+                Expr a = reconstruct_summation(gathered.a, op->type);
+                Expr b = reconstruct_summation(gathered.b, op->type);
+                return Select::make(op->condition, a, b) + like_terms;
+            }
+        } else {
+            return IRGraphMutator::visit(op);
+        }
+    }
+};
+
+// Used to bound the number of substitutions.
+class SubstituteSomeLets : public IRMutator {
```

Comment:
What does this do on something like: let b = a + a in let c = b + b in let d = c + c in ....

Does it produce an Expr of size 2 ^ count?

Reply:
Yes, I believe it does, the count only bounds the number of substtitutions, not the accumulated size. Would you prefer different behavior?

Reply:
Wait, something looks fishy here. You put values into the scope after mutating them, but then when you pull them out of the scope you mutate them a second time. I can't figure out what this would do with shadowed lets.

IMO we need to bound the actual size of the generated expression. 2^16 is still pretty damn big.

Reply:
I think the second mutation shouldn't happen, thanks for the catch.

Do you have any recommendations on how to bound overall size? I guess we could track multiplicative substitutions, i.e. in the example you gave, substituting `c` counts for 3 substitutions, as it has 2 `b` substitutions plus the actual `c` substutiton

Reply:
Are shadowed lets allowed in Halide syntax? `count_var_uses` definitely assumes that there are no shadowed lets. I thought #6583 was explicitly trying to prevent shadowed lets because we don't allow them.

Reply:
Conclusion from offline discussion: 

- This may be called in sliding window which is before removing shadowed lets. 
- Should be feasible to track total # substitutions by including an int in the Scope that counts the number of substitutions that occurred when mutating the RHS of that let

Reply:
Should we only substitute in the final expression?
i.e. suppose we have: `(let x = w + w in (let y = x + x in (let z = y + y in z + z)))` and we have substition count = 1. Should the resultant Expr be `(let x = w + w in (let y = (w + w) + x in (let z = y + y in z + z)))` or should it be `(let x = w + w in (let y = x + x in (let z = y + y in (y + y) + z)))` or neither? I assume (and prefer) the latter, but I don't think there's a good scheme to make that happen.

Also, if we have the count to do full substitutions, should the resultant Expr be:
`(let x = w + w in (let y = x + x in (let z = y + y in (((w + w) + (w + w)) + ((w + w) + (w + w))) + (((w + w) + (w + w)) + ((w + w) + (w + w))))))` or should the substitutions also happen inside of the let values? i.e. the Expr should become `(let x = w + w in (let y = (w + w) + (w + w) in ....`? I think this also ties into the above question - if we have a count of 1, and the scope for Exprs says that `z = ((w + w) + (w + w)) + ((w + w) + (w + w))` then we can't substitute that in because the tracked cost is too expensive, so perhaps the resultant Expr should be unchanged in this case?